### ISSUE TITLE
Very slow startup time on Windows

### ISSUE THREAD

**Issue by s-cerevisiae** [rktjmp/hotpot.nvim#55]
I was trying to port my Neovim config from my phone (with Termux) and Linux server to my Windows laptop, then I found it taking a lot of time to start on Windows: ~190ms on my phone, but ~2.7s on the Windows laptop, which is intolerable.

After some testing I confirmed that it's hotpot being slow.
`nvim --startuptime` showed that it takes 2456ms to load my `init.lua`. I manually compiled all my `.fnl`s to Lua, removed `require 'hotpot'` from `init.lua`, and the time reduced to 84ms.

I tried to make a minimal reproducible config, and here's what I got:

> init.lua:
> ```lua
> local function ensure_installed(plugin)
>   local author, name = unpack(vim.split(plugin, '/'))
>   local plugin_path = vim.fn.stdpath('data')..'/site/pack/packer/start/'..name
>   if vim.fn.empty(vim.fn.glob(plugin_path)) ~= 0 then
>     print("Installing", plugin, "to", plugin_path)
>     vim.fn.system({'git', 'clone', 'https://github.com/'..plugin, plugin_path})
>   end
> end
> ensure_installed('rktjmp/hotpot.nvim')
> require 'hotpot'
> require 'start'
> ```

> fnl/start.fnl
> ```fennel
> (print "do something")
> ```

This config can be used on any machine with a fresh install of Neovim.
In a Linux container it loads `init.lua` in 1ms and `hotpot.nvim` in 0.1ms after the second startup,
while on (another) Windows laptop loading `init.lua` always takes ~200ms. (hotpot takes about 0.2ms).
It also seems that the `fennel` module is always loaded on Windows.

I have 2 Windows devices and am willing to provide any additional information, try any solutions or do any tests as requested.

**Comment by rktjmp** [comment]
Interesting! I'm pretty busy atm so I'm not sure when I can dive into debugging it but maybe I can push you in the right directions.

I would assume if fennel is loading every time, hotpot is trying compile itself every time too. You could actually try this branch, see if it has any effect, which adjusts the compile target if needed: https://github.com/rktjmp/hotpot.nvim/tree/nix-compat

As for the 2.7s start up time, not really sure why that might be happening, either scanning some directories is really slow, or maybe fennel is never making it into `package.loaded` and is being re-required entirely each time? No idea why it wouldn't be cached though.

You could try changing `hotpot.lua#100` from
```lua
 local hotpot_dir = string.match((debug.getinfo(1, "S")).source, "@(.+)..?lua..?hotpot%.lua$")
```

to

```lua
 local hotpot_dir = "c:\\your\\path\\hotpot.nvim\\" -- check path sep is correct though
```

Just to be sure the debug call isn't super slow on windows (it wasn't when I tested it but that was in a VM on Win11, may be different in different setups).

Additional you have access to [`uv` and `uv.hrtime`](https://github.com/luvit/luv/blob/master/docs.md#uvhrtime) in that file, so you could try putting some `start_time` `stop_time` debugging around the `compile_dir` and `compile_string` function calls? 

Perhaps you could `vim.inspect(package.loaded["fennel"])` inside around `compile_string` too, just to make sure fennel isn't being loaded from disk every time (seems *super unlikely*, otherwise all of lua would feel really slow).

Probably run a lua-formatter on that file to make it a *bit* more readable.

If you want to work with fennel code instead you can install fennel and run `./build watch`.

Re the docs, most package managers will automatically index help files but since you're manually grabbing it you need to do it yourself, see the (newish) install instructions in the readme.
```lua
if vim.fn.empty(vim.fn.glob(hotpot_path)) > 0 then
  print("Could not find hotpot.nvim, cloning new copy to", hotpot_path)
  vim.fn.system({'git', 'clone',
                 'https://github.com/rktjmp/hotpot.nvim', hotpot_path})
  vim.cmd("helptags " .. hotpot_path .. "/doc")
end
```

**Comment by s-cerevisiae** [comment]
Well it seems that the docs problem is from `packer.nvim`. It isn't even indexing the docs for itself on Windows...
Maybe I'm going to open another issue there.

> You could try changing `hotpot.lua#100` from
> 
> ```lua
>  local hotpot_dir = string.match((debug.getinfo(1, "S")).source, "@(.+)..?lua..?hotpot%.lua$")
> ```
> 
> to
> 
> ```lua
>  local hotpot_dir = "c:\\your\\path\\hotpot.nvim\\" -- check path sep is correct though
> ```

This makes no observable difference. I'm going to try `hrtime`.

**Comment by s-cerevisiae** [comment]
These are what I've observed till now:

Firstly it seems that symlink to `canary` is never created: symlinks on windows require admin rights to be created (which is absurd), so hotpot gets compiled every time. Running `nvim` with admin rights gets rid of 200ms from self-compilation, so the minimal example would be starting immediately. For my personal config there's still 1400ms to go though.

Secondly I checked that `hotpot.fennel` is cached in `package.loaded` so it's probably not getting loaded again and again.

Thirdly, `compile-file` is called for each `.fnl` file each time nvim is started.

**Comment by rktjmp** [comment]
When you say admin rights, does that mean "admin account" or "elevated privileges"? I assume the VM I used had a default "admin account" but I never needed to go one step up to root-like via those pop-up boxes. I assume you're using an actual "user account"?

Can't imagine where that 1.4s comes from right now, once hotpot is compiled it doesn't really do anything extra over regular nvim besides checking for ".fnl" files next to ".lua" files, not 1.4s extra work unless those checks are maybe hitting another permission error and logging that error is grinding.

**Comment by s-cerevisiae** [comment]
> When you say admin rights, does that mean "admin account" or "elevated privileges"?

Sorry for being unclear, I meant elevated priviledges. And after some googling I found another way: after enabling "Developer Mode" in settings, apps can create symlinks without elevated priviledge. Despite that the description of this option says nothing about it.
Tested, and it actually works.

> Can't imagine where that 1.4s comes from right now, once hotpot is compiled it doesn't really do anything extra over regular nvim besides checking for ".fnl" files next to ".lua" files, not 1.4s extra work unless those checks are maybe hitting another permission error and logging that error is grinding.

Compiled lua files in the temp folder are recompiled everytime I run nvim. I guess there's something wrong with the check.

**Comment by rktjmp** [comment]
Wonder if windows doesn't return a consistent modified time? Or not a posix int (assumed libuv was supposed to x-platform that but maybe not)?

Basically we have this code:

https://github.com/rktjmp/hotpot.nvim/blob/8ccf600ccb5dbc28ea329d641a58e2fd6974d38e/fnl/hotpot/index.fnl#L88-L100

where `files` is a list of related files to a module (normally just the source but may be macro files too), `timestamp` is the time we compiled at and `loader` is the loader we created at that time. 

The timestamp is just gotten from `uv.stat`,

https://github.com/rktjmp/hotpot.nvim/blob/8ccf600ccb5dbc28ea329d641a58e2fd6974d38e/fnl/hotpot/fs.fnl#L34-L38

The line numbers in the Fennel code should match up directly with the lines in the (unformatted...) lua code if you're working that way.

Are you able to poke around in there a bit, see if the mtimes are somehow always getting stale or something?

**Comment by s-cerevisiae** [comment]
The msgpack data in `index.bin` is somehow corrupted. I tried reading it manually, and nvim gives `E5108: Error executing lua [string ":lua"]:1: incomplete msgpack string`.

I checked the size of `index.bin` file, it's 2094KB on my phone but 1441KB and rarely 635KB on my laptop...There might be something weird happening with `dehydrate-records`.

**Comment by rktjmp** [comment]
Does impatient.nvim work on windows? We use basically the same idea from there, maybe it's some kind of encoding issue. If that's broken, I could imagine trying to load it over and over would give some slow down or something. Again sorry I can't get in the weeds of it right now (also not sure I could reproduce it tbh), your feedback is very helpful.

**Comment by s-cerevisiae** [comment]
Ah I found something [here](https://www.lua.org/pil/21.2.2.html)...

> The simple model functions io.input and io.output always open a file in text mode (the default). In Unix, there is no difference between binary files and text files. But in some systems, notably Windows, binary files must be opened with a special flag. To handle such binary files, you must use io.open, with the letter `b´ in the mode string. 

Replaced "r" with "rb" and "w" with "wb", now it's ~200ms total.
SOLVED.

**Comment by s-cerevisiae** [comment]
Conclusion:
1. For Windows users, "Developer Mode" should be enabled in order to create the symbol link.
2. `index.bin` should be opened with mode string "rb" and "wb" for reads and writes respectively.

@rktjmp What do you think about the symlink usage? Should it be documented (for that "Developer Mode" thing) or be avoided and track the canary in another way (like a plain text file containing the hash string)?

**Comment by rktjmp** [comment]
I've marked you as the author of the fix if that is ok with you:

https://github.com/rktjmp/hotpot.nvim/compare/fix-windows-index

Currently we use a symlink because I figured `stat -> compare int`ing a file is (fractionally) faster than `stat -> open -> read -> compare string` and the file just "disappearing"  (when the symlink becomes invalid) is a good way to catch all "needs recompile" cases.

It may change in the future but enabling developer mode on windows to use what is for the most part, a developer tool doesn't seem that much of a burden?

I bet the MS VM image I got already had that enabled already, as it came with VSCode installed and a compiler toolchain, which is probably why it never came up before.

**Comment by s-cerevisiae** [comment]
Well I think the documentation must be extra clear about we are only using the symlink feature, because turning on that Developer Mode switch shows up a big freaking warning like "this switch enables installing and running arbitrary (uwp) apps from outside ms store, use it at your own risk", which sounds super suspicious.

Also the link should point [here](https://blogs.windows.com/windowsdeveloper/2016/12/02/symlinks-windows-10/), because the symlink permission thing is mentioned in absolutely nowhere else...

Otherwise it seems ok for me.

**Comment by rktjmp** [comment]
Fixed in 5bed27726898bff5cd82b370b8a07b5ed7ebb50b  bfa18a6a03f61c91e58627f92e72f62a122c02dd.

Thanks so much for your work.
